File: ios/Shared/Analytics.swift
import Foundation

func averageReturn(_ history: [DayData], days: Int? = nil) -> Double {
    let slice = days != nil ? Array(history.suffix(days!)) : history
    guard !slice.isEmpty else { return 0 }
    return slice.reduce(0.0) { $0 + $1.totalReturn } / Double(slice.count)
}

func bestDay(_ history: [DayData]) -> DayData? {
    history.max(by: { $0.totalReturn < $1.totalReturn })
}

func longestStreak(_ history: [DayData]) -> Int {
    history.map(\.streak).max() ?? 0
}

func habitConsistency(_ history: [DayData], habitId: String) -> Double {
    guard !history.isEmpty else { return 0 }
    let daysUsed = history.filter { day in
        let allocs = blocksToAllocations(day.blocks)
        return (allocs[habitId] ?? 0) > 0
    }.count
    return Double(daysUsed) / Double(history.count) * 100
}

func categoryDistribution(_ history: [DayData], allHabits: [Habit]) -> (essential: Double, growth: Double, drain: Double) {
    guard !history.isEmpty else { return (0, 0, 0) }

    var totalBlocks = 0
    var essentialBlocks = 0
    var growthBlocks = 0
    var drainBlocks = 0

    for day in history {
        let allocs = blocksToAllocations(day.blocks)
        for (habitId, count) in allocs {
            if let habit = habitById(habitId, from: allHabits) {
                totalBlocks += count
                switch habit.category {
                case .essential: essentialBlocks += count
                case .growth: growthBlocks += count
                case .drain: drainBlocks += count
                }
            }
        }
    }

    guard totalBlocks > 0 else { return (0, 0, 0) }

    return (
        essential: Double(essentialBlocks) / Double(totalBlocks) * 100,
        growth: Double(growthBlocks) / Double(totalBlocks) * 100,
        drain: Double(drainBlocks) / Double(totalBlocks) * 100
    )
}

func dailyReturnsData(_ history: [DayData]) -> [(date: String, totalReturn: Double)] {
    history.map { (date: $0.date, totalReturn: $0.totalReturn) }
}

struct TopHabitEntry: Identifiable {
    let id: String
    let habit: Habit
    let totalBlocks: Int
    let avgBlocks: Double
}

func topHabits(_ history: [DayData], allHabits: [Habit], limit: Int = 5) -> [TopHabitEntry] {
    guard !history.isEmpty else { return [] }

    var habitTotals: [String: Int] = [:]

    for day in history {
        let allocs = blocksToAllocations(day.blocks)
        for (habitId, count) in allocs {
            habitTotals[habitId, default: 0] += count
        }
    }

    return habitTotals
        .compactMap { (habitId, total) -> TopHabitEntry? in
            guard let habit = habitById(habitId, from: allHabits) else { return nil }
            return TopHabitEntry(
                id: habitId,
                habit: habit,
                totalBlocks: total,
                avgBlocks: Double(total) / Double(history.count)
            )
        }
        .sorted { $0.totalBlocks > $1.totalBlocks }
        .prefix(limit)
        .map { $0 }
}

File: src/lib/analytics.ts
import { DayData, Habit } from './habits'
import { blocksToAllocations } from './calculations'

/** Calculate average return over the last N days */
export function averageReturn(history: DayData[], days?: number): number {
  const slice = days ? history.slice(-days) : history
  if (slice.length === 0) return 0
  return slice.reduce((sum, d) => sum + d.totalReturn, 0) / slice.length
}

/** Find the best performing day */
export function bestDay(history: DayData[]): DayData | null {
  if (history.length === 0) return null
  return history.reduce((best, d) => d.totalReturn > best.totalReturn ? d : best)
}

/** Find the longest streak in history */
export function longestStreak(history: DayData[]): number {
  if (history.length === 0) return 0
  return Math.max(...history.map(d => d.streak))
}

/** Calculate what percentage of days a habit was used */
export function habitConsistency(history: DayData[], habitId: string): number {
  if (history.length === 0) return 0
  const daysUsed = history.filter(d => {
    const allocs = blocksToAllocations(d.blocks)
    return (allocs[habitId] || 0) > 0
  }).length
  return (daysUsed / history.length) * 100
}

/** Get category distribution as percentages */
export function categoryDistribution(
  history: DayData[],
  allHabits: readonly Habit[]
): { essential: number; growth: number; drain: number } {
  if (history.length === 0) return { essential: 0, growth: 0, drain: 0 }

  let totalBlocks = 0
  const categoryBlocks = { essential: 0, growth: 0, drain: 0 }

  history.forEach(day => {
    const allocs = blocksToAllocations(day.blocks)
    Object.entries(allocs).forEach(([habitId, count]) => {
      const habit = allHabits.find(h => h.id === habitId)
      if (habit) {
        categoryBlocks[habit.category] += count
        totalBlocks += count
      }
    })
  })

  if (totalBlocks === 0) return { essential: 0, growth: 0, drain: 0 }

  return {
    essential: (categoryBlocks.essential / totalBlocks) * 100,
    growth: (categoryBlocks.growth / totalBlocks) * 100,
    drain: (categoryBlocks.drain / totalBlocks) * 100,
  }
}

/** Get daily returns */
export function dailyReturnsData(history: DayData[]): { date: string; totalReturn: number }[] {
  return history.map(d => ({
    date: d.date,
    totalReturn: d.totalReturn,
  }))
}

/** Get the top habits by time spent across history */
export function topHabits(
  history: DayData[],
  allHabits: readonly Habit[],
  limit = 5
): { habit: Habit; totalBlocks: number; avgBlocks: number }[] {
  if (history.length === 0) return []

  const habitTotals: Record<string, number> = {}

  history.forEach(day => {
    const allocs = blocksToAllocations(day.blocks)
    Object.entries(allocs).forEach(([habitId, count]) => {
      habitTotals[habitId] = (habitTotals[habitId] || 0) + count
    })
  })

  return Object.entries(habitTotals)
    .map(([habitId, total]) => {
      const habit = allHabits.find(h => h.id === habitId)
      if (!habit) return null
      return { habit, totalBlocks: total, avgBlocks: total / history.length }
    })
    .filter((item): item is NonNullable<typeof item> => item !== null)
    .sort((a, b) => b.totalBlocks - a.totalBlocks)
    .slice(0, limit)
}
